formulas/glance/configure.sls

```diff
@@ -74,3 +74,66 @@ glance_api_service:
     - enable: true
     - watch:
       - file: /etc/glance/glance-api.conf
+
+/etc/openstack/clouds.yml:
+  file.managed:
+    - source: salt://formulas/common/openstack/files/clouds.yml
+    - makedirs: True
+    - template: jinja
+    - defaults:
+        password: {{ pillar['openstack']['admin_password'] }}
+        auth_url: {{ constructor.endpoint_url_constructor(project='keystone', service='keystone', endpoint='internal') }}
+
+{% if not salt['file.directory_exists' ]('/kvm') %}
+kvm_dir:
+  file.directory:
+    - name: /kvm
+{% endif %}
+
+{% if not salt['file.directory_exists' ]('/kvm/glance_images') %}
+/kvm/glance_images:
+  file.directory:
+    - makedirs: True
+{% endif %}
+
+{% if not salt['file.directory_exists' ]('/kvm/glance_templates') %}
+/kvm/glance_templates:
+  file.directory:
+    - makedirs: True
+{% endif %}
+
+{% if grains['os_family'] == 'RedHat' %}
+libvirtd_service:
+  service.running:
+    - name: libvirtd
+    - enable: true
+{% endif %}
+
+{% for os, args in pillar.get('glance_images', {}).items() %}
+/kvm/glance_templates/{{ args['image_name'] }}.yaml:
+  file.managed:
+    - template: jinja
+    - contents: |
+        image_name: {{ args.get('image_name', '') }}
+        method: {{ args.get('method', '') }}
+        image_url: {{ args.get('image_url', '') }}
+        image_size: {{ args.get('size', '')}}
+        conversion: {{ args.get('conversion', '') }}
+        input_format: {{ args.get('input_format', '') }}
+        output_format: {{ args.get('output_format', '') }}
+        packages: {{ args.get('packages', '') }}
+        customization: |
+            {{ args.get('customization', '') | indent(12) }}
+
+create_glance_image_{{ args['image_name'] }}:
+  cmd.run:
+    - name: 'python3 /tmp/image_bakery/image_bake.py -t /kvm/glance_templates/{{ args['image_name']}}.yaml -o /kvm/glance_images'
+    - onchanges: [ /kvm/glance_templates/{{ args['image_name'] }}.yaml ]
+
+upload_glance_image_{{ args['image_name'] }}:
+  glance_image.present:
+    - name: {{ args.get('image_name') }}
+    - onchanges: [ /kvm/glance_templates/{{ args['image_name'] }}.yaml ]
+    - filename: '/kvm/glance_images/{{ args.get('image_name') }}'
+    - image_format: {{ args.get('output_format') }}
```

Comment:
Currently this would run on every glance vm and anytime the yamls changed. Should it be separated out into a new sls and only ran if needed? I think it should be in a spawn 0 to eliminate multiple uploads of the same image. The other thing is that it has no logic for if the image already exists, do something with the old image before uploading a new one as it will cause image name conflicts if it already exists.

Reply:
I missed a couple lines in that snippet, starts at 133

Reply:
Actually I may be wrong  with use of glance_image.present. I should have read the docs first.

Reply:
on consecutive runs after changing the yaml it does not appear to upload a new image to openstack, will do some more testing